src/cursor/abstract_cursor.ts

```diff
@@ -0,0 +1,710 @@
+    return maybePromise(callback, done => {
+      if (this[kId] === Long.ZERO) {
+        return done(undefined, false);
+      }
+
+      if (this[kDocuments].length) {
+        return done(undefined, true);
+      }
+
+      next(this, (err, doc) => {
+        if (err) return done(err);
+
+        if (doc) {
+          this[kDocuments].unshift(doc);
+          done(undefined, true);
+          return;
+        }
+
+        done(undefined, false);
+      });
+    });
+  }
+
+  /** Get the next available document from the cursor, returns null if no more documents are available. */
+  next(): Promise<Document | null>;
+  next(callback: Callback<Document | null>): void;
+  next(callback?: Callback<Document | null>): Promise<Document | null> | void {
+    return maybePromise(callback, done => {
+      if (this[kId] === Long.ZERO) {
+        return done(new MongoError('Cursor is exhausted'));
+      }
+
+      next(this, done);
+    });
+  }
+
+  /**
+   * Iterates over all the documents for this cursor using the iterator, callback pattern.
+   *
+   * @param iterator - The iteration callback.
+   * @param callback - The end callback.
+   */
+  forEach(iterator: (doc: Document) => boolean | void): Promise<void>;
+  forEach(iterator: (doc: Document) => boolean | void, callback: Callback<void>): void;
+  forEach(
+    iterator: (doc: Document) => boolean | void,
+    callback?: Callback<void>
+  ): Promise<void> | void {
+    if (typeof iterator !== 'function') {
+      throw new TypeError('Missing required parameter `iterator`');
+    }
+
+    return maybePromise(callback, done => {
+      const transform = this[kTransform];
+      const fetchDocs = () => {
+        next(this, (err, doc) => {
+          if (err || doc == null) return done(err);
+          if (doc == null) return done();
+
+          // NOTE: no need to transform because `next` will do this automatically
+          let result = iterator(doc);
+          if (result === false) return done();
+
+          // these do need to be transformed since they are copying the rest of the batch
+          const internalDocs = this[kDocuments].splice(0, this[kDocuments].length);
+          if (internalDocs) {
+            for (let i = 0; i < internalDocs.length; ++i) {
+              result = iterator(transform ? transform(internalDocs[i]) : internalDocs[i]);
+              if (result === false) return done();
+            }
+          }
+
+          fetchDocs();
+        });
+      };
+
+      fetchDocs();
+    });
+  }
+
+  close(): void;
+  close(callback: Callback): void;
+  close(options: CursorCloseOptions): Promise<void>;
+  close(options: CursorCloseOptions, callback: Callback): void;
+  close(options?: CursorCloseOptions | Callback, callback?: Callback): Promise<void> | void {
+    if (typeof options === 'function') (callback = options), (options = {});
+    options = options || {};
+
+    const needsToEmitClosed = !this[kClosed];
```

Comment:
Seems this is only used once and could be moved down into the lower `if` check, maybe with a comment `// needs to emit close`?